pkg/core/combat/shape.go

```diff
@@ -1,131 +1,117 @@
-	return sq <= math.Pow(c.r, 2)
+	return false
 }
 
-func (c *Circle) Pos() (float64, float64) {
-	return c.x, c.y
-}
+// shared between Circle and Rectangle
+// https://stackoverflow.com/questions/401847/circle-rectangle-collision-detection-intersection
+// https://yal.cc/rot-rect-vs-circle-intersection/
+func IntersectRectangle(r Rectangle, c Circle) bool {
+	// TODO: rectangle-circle with fanAngle hurtbox/hitbox collision
+	if c.segments != nil {
+		panic("fanAngle hitbox and hurtbox aren't supported in rectangle-circle collision")
+	}
 
-func (c *Circle) SetPos(x, y float64) {
-	c.x = x
-	c.y = y
-}
+	// AABB test
+	if !AABBTest(r.aabb, c.aabb) {
+		return false
+	}
 
-type Rectangle struct {
-	x, y, w, h float64
-}
+	// set origin to rectangle center by shifting circle center position
+	relative := c.center.Sub(r.center)
 
-func (r *Rectangle) String() string {
-	return fmt.Sprintf("w: %v h: %v x: %v y: %v", r.w, r.h, r.x, r.y)
-}
+	// take direction from rectangle and rotate circle center in the opposite direction to remove rectangle rotation
+	dir := r.dir.Mul(Point{X: -1, Y: 1})
+	local := relative.Rotate(dir)
 
-func (r *Rectangle) IntersectCircle(c Circle) bool {
-	cdX := math.Abs(c.x - r.x)
-	cdY := math.Abs(c.y - r.y)
+	// constrain circle center to one quadrant
+	local.X = math.Abs(local.X)
+	local.Y = math.Abs(local.Y)
 
-	if cdX > (r.w/2 + c.r) {
-		return false
+	topRight := Point{
+		X: r.w / 2,
+		Y: r.h / 2,
 	}
-	if cdY > (r.h/2 + c.r) {
+
+	// eliminate cases in which the circle center is too far away from rectangle edges
+	if local.X > c.r+topRight.X || local.Y > c.r+topRight.Y {
 		return false
 	}
-	if cdX <= (r.w / 2) {
-		return true
-	}
-	if cdY <= (r.h / 2) {
+
+	// circle center has to be close enough to the rectangle edges at this point
+	// -> accept if circle center is within the coordinate area 0 <= x <= r.w/2 || 0 <= y <= r.h/2
+	// -> if it's in that area, then it definitely intersects with one edge
+	if local.X <= topRight.X || local.Y <= topRight.Y {
 		return true
 	}
-	sq := math.Pow(cdX-r.w/2, 2) + math.Pow(cdY-r.h/2, 2)
-	return sq <= math.Pow(c.r, 2)
-}
-
-func (r *Rectangle) IntersectRectangle(r2 Rectangle) bool {
-	halfr2w := r2.w / 2
-	halfr2h := r2.h / 2
-	halfr1w := r.w / 2
-	halfr1h := r.h / 2
-	return r.x+halfr1w >= r2.x-halfr2w && //right side >= r2 left side
-		r.x-halfr1w <= r2.x+halfr2w && //left side <= r2 right side
-		r.y+halfr1h >= r2.y-halfr2h && //top side  >= r2 bot side
-		r.y-halfr1h <= r2.y+halfr2h //bot side >= r2 topside
-}
 
-func (r *Rectangle) Pos() (float64, float64) {
-	return r.x, r.y
-}
+	// circle center is in the area r.w/2 < x <= r.w/2+radius && r.h/2 < y <= r.h/2+radius
+	// -> it can only intersect if it's close enough to the topRight corner
+	if local.Sub(topRight).MagnitudeSquared() > c.r*c.r {
+		return false
+	}
 
-func (r *Rectangle) SetPos(x, y float64) {
-	r.x = x
-	r.y = y
+	// intersection is guaranteed at this point
+	return true
```

Comment:
This is inconsistent with circle-circle rect-rect collision. In circle-circle and rect-rect collision we treat the `==` case as `not-colliding`/`false` whereas here we treat it as `colliding`/`true`. I think either approach is fine, just need to make sure all collision checks treat the equals case the same (not sure what unity does).

Reply:
Gonna go with `==` meaning `colliding`/`true`.